### ISSUE TITLE
Allow to use subscript numbers in F# code

### ISSUE THREAD

**Issue by ForNeVeR** [fsharp/fslang-suggestions#1272]
I sometimes work with scientific code and it's inconvenient in some cases. When I program a formula, it is very convenient to enter it as is. For example, there's a formula `Z₀ = E / H` (the first real example I found in my notes when performing calculations related to the [impedance of free space](https://en.wikipedia.org/wiki/Impedance_of_free_space)), which I'd like to see in my code as it looks here. But unfortunately, in F# I have to write it as `Z_0 = E / H`, since `Z₀` doesn't compile ([link to Sharplab](https://sharplab.io/#v2:EYLgtghglgdgNAGxAMwM5wC4iguATEAagB8B7ABwFMYACAZQE9UNKwBYAKAUoxoFEaAXhoBGAAyduvABJCaAVglceNAFqAAgjkCA9DWmcgA=)).

> ```
> error FS0010: Unexpected character '₀' in binding. Expected '=' or other token.
> ```

**I propose we** add a possibility to use subscript numbers in F# code, such as:
```fsharp
let E = 10
let H = 50

let Z₀ = E / H // this!
```

Notably, it's possible to use subscript characters such as `ₖₒₐₗₐ`, but not numbers.

**The existing way of approaching this problem in F# is** to use a notation of `Z_0` instead, which doesn't look that good.

## Pros and Cons 

**The advantages of making this adjustment to F# are**: there will be easier to write and, more importantly, read code based on formulas in scientific calculations.

**The disadvantages of making this adjustment to F# are**: well, it'd be easier to abuse this feature and name certain symbols in 
- either a confusing way
- or in a way that makes it harder to reuse the values in other languages (say, C# doesn't allow `Z₀` either).

Note that this is about subscript numbers for now, and not superscripts. Superscripts would be more confusing, since `E²` is easy to mix with `E ** 2`.

## Extra information

**Estimated cost (XS, S, M, L, XL, XXL):**  XS.

**Related suggestions:** none that I was able to find.

## Affidavit (please submit!)

Please tick this by placing a cross in the box:
* [x] This is not a question (e.g. like one you might ask on [stackoverflow](http://stackoverflow.com)) and I have searched stackoverflow for discussions of this issue
* [x] I have [searched both open and closed suggestions on this site](http://github.com/fsharp/fslang-suggestions/issues) and believe this is not a duplicate
* [x] This is not something which has obviously "already been decided" in previous versions of F#.  If you're questioning a fundamental design decision that has obviously already been taken (e.g. "Make F# untyped") then please don't submit it.

Please tick all that apply:
* [x] This is not a breaking change to the F# language design
* [x] I or my company would be willing to help implement and/or test this

## For Readers

If you would like to see this issue implemented, please click the :+1: emoji on this issue. These counts are used to generally order the suggestions by engagement.

**Comment by vzarytovskii** [comment]
I guess it's a similar area as https://github.com/fsharp/fslang-suggestions/issues/1104 and (probably) https://github.com/fsharp/fslang-suggestions/issues/1079. Former was rejected, latter is still open.

cc @dsyme

**Comment by vzarytovskii** [comment]
My main problem with those, personally, is that they're really annoying to type on normal keyboards (even with `Win+;` and similar).

**Comment by ForNeVeR** [comment]
`ₖₒₐₗₐ` is also annoying but still allowed.

**Comment by dsyme** [comment]
I have no idea how to type any of this stuff on a keyboard in standard editors (say VSCode to start).  If there are simple standard input methods that don't involve copy-and-paste this stuff starts to make much more sense.

That said I'm really surprised that subscript-o is not being considered a valid number-in-identifier character.  The F# spec aligns with the C# spec here and allows "subscript letters" (e.g. [subscript letter o](https://www.compart.com/en/unicode/U+2092#:~:text=%E2%80%9C%E2%82%92%E2%80%9D%20U%2B2092%20Latin%20Subscript%20Small%20Letter%20O%20Unicode%20Character)) which are Unicode category `Lm`.   

For digits it only allows `[0-9]`.  This actually surprises me as it means these aren't allowed:
* [All the decimal numbers](https://www.compart.com/en/unicode/category/Nd)
* [All the letter numbers](https://www.compart.com/en/unicode/category/Nl)
* [All the other numbers](https://www.compart.com/en/unicode/category/No)

Now I mean I can see plenty of problems with a whole lot of these - e.g. one of the "other numbers" is ⒀ and should 

```fsharp
let ⒀⒀⒀⒀⒀ = 1 
```

But then the vast number of [other letters](https://www.compart.com/en/unicode/category/Lo) haven't all been vetted for being 100% sensible in code identifiers.


My gut feeling is that it would be reasonable, consistent and acceptable to adjust the F# spec to allow `Nd`, `Nl`, `No` in identifiers.  The current situtation where subscript letters are allowed but not subscript digits feels wrong, and admitting these numeric categories into identifiers seems the most reasonable and consistent solution to this.

**Comment by dsyme** [comment]
Here's an example of subscript-letter-o in an F# identifier: https://sharplab.io/#v2:EYLgtghglgdgNAGxAMwM5wC4iguATEAagB8B7ABwFMYACAZQE9UNKwBYAKAUoxoFEaAXhoBGAAyduvABJCaAVglceNAFqAkgjkCA9DWmcgA=

**Comment by zanaptak** [comment]
I think it's reasonable to consider specific characters that are highly beneficial for readability rather than whole character categories. In fact this idea is discussed in Unicode Standard Annex 31 under [Default Identifier Syntax](https://unicode.org/reports/tr31/#Default_Identifier_Syntax) where they suggest allowing the set of:

> [⁽₍⁾₎⁺₊⁼₌⁻₋⁰₀¹₁²₂³₃⁴₄⁵₅⁶₆⁷₇⁸₈⁹₉]

rather than the full Other_Number category.

Regarding difficulty of typing, in the domains where this has the most value such as scientific and mathematical computing, those developers are likely well-versed in how to input them, and there's autocompletion and copy-paste for the rest of us.

**Comment by ForNeVeR** [comment]
Regarding typing: I personally use [WinCompose](https://github.com/samhocevar/wincompose) and press `Meta` `_` `0` to print `₀` (notably, I used to do this while working on the same scientific project), though I of course understand that there are no widely-accepted means to type this across all (or even any of) the major platforms supported by F#.

**Comment by vzarytovskii** [comment]
When working on RFC, I would appreciate the inclusion of the tooling aspect, such as - how does it work on different OSs and environments, both typing those as well as displaying - vim/Emacs in terminal, popular gui editors and IDEs, github diffs and markdown highlighting, etc.

**Comment by vzarytovskii** [comment]
> @vzarytovskii I get the need, though I think all those issues apply for F# 1.0 onwards, and C# 1.0 too - all the unicode characters available via this huge set of unicode characters: 
> 
> ```
> regexp letter-char = '\Lu' | '\Ll' | '\Lt' | '\Lm' | '\Lo' | '\Nl'
> ```
> 
> So this isn't really raising anything new - it's just proposing a relatively modest revision to the existing spec of identifiers, to take into account Annex 31, which proibably wasn't written at the time of C# 1.0
>  

Yeah, I'm just curious of the level of support for other tooling, mostly because github client on my phone and tablet render it as just squares :)

**Comment by dsyme** [comment]
> Yeah, I'm just curious of the level of support for other tooling, mostly because github client on my phone and tablet render it as just squares :)

Might be good to open a separate language/tooling suggestion - discussing the state of input method entry for unicode characters, especially ones that might reasonably be used in coding, like the examples above. 

I guess it's a topic that tends to progress and change over time

**Comment by zanaptak** [comment]
Despite my earlier comment, I would probably lean toward the full-category approach for F# as a general purpose language. It avoids the subjectivity and potential user confusion around which characters get first-class treatment.

**Comment by charlesroddie** [comment]
For Unicode conformance, F# can specify a profile of characters to add or subtract to XID_Start and XID_Continue. As discussed above, there is a proposal to add a math profile and then F# could just take that.
- Unicode math profile: https://www.unicode.org/L2/L2022/22230-math-profile.pdf
- Clang implementation of draft: https://reviews.llvm.org/D137051 (They did this fast likely because these characters were de-facto allowed before.)
- Current Unicode 15.1.0 **draft** 7 which contains the new profile recommendations. https://www.unicode.org/reports/tr31/tr31-38.html

Current CSharp and IL (I just checked in SharpLab) doesn't allow these characters in identifiers. CSharp currently doesn't have a very clear definition of allowed characters beyond what is allowed in practice now, or what was defined in ECMA-330 in Jan 2000. https://github.com/dotnet/csharpstandard/issues/305 looks at resolving this. The cleanest path would be if this were proposed there.

If this is implemented in F# without dotnet support, then the IL generated would need to be manipulated or escaped. A spec for this might be to treat the following as equivalent:

```
let x₀ = 
let ``x₀`` =
```

From SharpLab, this goes into IL as `'x_0'` and doesn't have a corresponding C# syntax.

**Comment by xp44mm** [comment]
Subscripts are suitable for writing chemical Molecular formula, which is very intuitive. For example, 

```fsharp
if x = Mole.H₂O then 
...
```
